Allow arbitrary bond yield solver

TheSpiritXIII opened the pull request:
This PR adds a template parameter to CashFlows and BondFunctions to allow them to take an arbitrary solver, instead of the default NewtonSafe solver (such as the Brent solver). These changes are backwards compatible due to the change involving a function template overload.

I am willing to make any requested further changes and can squash them if requested.

lballabio commented on ql/pricingengines/bond/bondfunctions.cpp:
```diff
@@ -369,20 +368,8 @@ namespace QuantLib {
-                                dayCounter, compounding, frequency,
-                                false, settlement, settlement,
-                                accuracy, maxIterations, guess);
+        yield<NewtonSafe>(bond, cleanPrice, dayCounter, compounding, frequency,
```
This should return the value, not just call `yield`.

lballabio commented on ql/pricingengines/bond/bondfunctions.hpp:
```diff
@@ -153,6 +155,32 @@ namespace QuantLib {
                           Size maxIterations = 100,
                           Rate guess = 0.05);
+        template <typename Solver>
+        static Rate yield(const Bond& bond,
```
I think this (and also `CashFlows::yield`) should take the solver as an argument.  This way, it would be possible to set other parameters to the solver, as well as to use solvers that don't have a default constructor.

TheSpiritXIII commented:
Hello @lballabio, I made and squashed the changes you requested and rebased my commit on top of the latest master. Let me know if there are any further changes you want me to make.
